### PR TITLE
Animation + Domain + texture perfs + STL orientation + ImageMapper slicing

### PR DESCRIPTION
**Pull request by agirault**

### CODE REVIEW

File: Sources/macro.js [hunk 1]
```diff
@@ -1150,6 +1152,10 @@ export function proxy(publicAPI, model) {
   publicAPI.getPropertyByName = (name) =>
     getProperties().find((p) => p.name === name);
 
+  publicAPI.getPropertyDomainByName = (name) => {
+    return propertyMap[name].domain;
+  };
+
```
Comment (jourdain) [inline Sources/macro.js]: need to be reformated

`publicAPI.getPropertyDomainByName = (name) => propertyMap[name].domain;`

File: Sources/IO/Geometry/STLReader/index.js [hunk 2]
```diff
@@ -77,6 +93,15 @@ function vtkSTLReader(publicAPI, model) {
     model.parseData = content;
 
     // Binary parsing
+    // Header
+    const headerData = new Uint8Array(content, 0, 80);
+    let headerStr = '';
+    for (let i = 0; i < 80; ++i) {
+      headerStr += String.fromCharCode(headerData[i]);
+    }
+    const header = parseHeader(headerStr);
```
Comment (jourdain) [inline Sources/IO/Geometry/STLReader/index.js]: There is a better way to do it. See the XML readers.

File: Sources/IO/Geometry/STLReader/index.js [hunk 3]
```diff
@@ -110,6 +135,49 @@ function vtkSTLReader(publicAPI, model) {
       cellDataValues[faceIdx] = dataView.getUint16(offset + 48, true);
     }
 
+    // Rotate points
+    const orientationField = 'SPACE';
+    if (orientationField in header && header[orientationField] !== 'LPS') {
+      const XYZ = header[orientationField];
+      const mat4 = [...new Array(16)].map((x) => 0);
```
Comment (jourdain) [inline Sources/IO/Geometry/STLReader/index.js]: new Float32Array(16); will already be filled with 0.

File: Sources/IO/Geometry/STLReader/index.js (hunk 2 above)
Comment (agirault) [reply Sources/IO/Geometry/STLReader/index.js]: this?
https://github.com/Kitware/vtk-js/blob/master/Sources/IO/XML/XMLReader/index.js#L68-L95

Comment (jourdain) [reply Sources/IO/Geometry/STLReader/index.js]: https://github.com/Kitware/vtk-js/blob/master/Sources/IO/Core/BinaryHelper/index.js#L7-L19

File: Sources/IO/Geometry/STLReader/index.js [hunk 4]
```diff
@@ -77,6 +96,12 @@ function vtkSTLReader(publicAPI, model) {
     model.parseData = content;
 
     // Binary parsing
+    // Header
+    const headerData = new Uint8Array(content, 0, 80);
+    const headerStr = BinaryHelper.arrayBufferToString(headerData);
```
Comment (agirault) [inline Sources/IO/Geometry/STLReader/index.js]: @jourdain this works, but should I initialize `headerData` as something else than a `Uint8Array`?

Comment (jourdain) [reply Sources/IO/Geometry/STLReader/index.js]: Check with @floryst

**Comment by jourdain** [comment]
You have linter/prettier issue

File: Sources/IO/Geometry/STLReader/index.js (hunk 4 above)
Comment (floryst) [reply Sources/IO/Geometry/STLReader/index.js]: `arrayBufferToString` should take in an ArrayBuffer, but it'll work with UInt8Arrays.

Note the fallback code in `arrayBufferToString` will make a copy of your input if it's UInt8Array.

Comment (agirault) [reply Sources/IO/Geometry/STLReader/index.js]: used `slice()` to provide ArrayBuffer of proper offset and length.

File: Sources/IO/Geometry/STLReader/index.js (hunk 2 above)
Comment (agirault) [reply Sources/IO/Geometry/STLReader/index.js]: done

**Comment by agirault** [comment]
@jourdain fixed.

@martinken the patch to skip scaling textures (ef2cb1f) makes the ImageMapper tests fail. Is that expected? See below:
screen shot 2018-02-22 at 3 19 37 pm

**Comment by martinken** [comment]
Pending the other fix (webgl error) we talked about this morning those images look fine to me. But I suspect we may need to have two sets of images or just a larger threshold. To me they look the same but I get that they are different because on one we are interpolating to a power of two while on the other opengl is using the npot image

File: Sources/Rendering/OpenGL/ImageMapper/index.js [hunk 5]
```diff
@@ -538,12 +538,13 @@ function vtkOpenGLImageMapper(publicAPI, model) {
...
       if (ijkMode === SlicingMode.I) {
-        scalars = [];
+        scalars = new basicScalars.constructor(dims[2] * dims[1]);
+        let id = 0;
         for (let k = 0; k < dims[2]; k++) {
           for (let j = 0; j < dims[1]; j++) {
-            scalars.push(
-              basicScalars[sliceOffset + j * dims[0] + k * dims[0] * dims[1]]
-            );
+            id = j * dims[2] + k;
```
Comment (floryst) [inline Sources/Rendering/OpenGL/ImageMapper/index.js]: Should this be `id = k * dims[1] + j`?

File: Sources/Rendering/OpenGL/ImageMapper/index.js [hunk 6]
```diff
@@ -561,12 +562,13 @@ function vtkOpenGLImageMapper(publicAPI, model) {
...
       } else if (ijkMode === SlicingMode.J) {
-        scalars = [];
+        scalars = new basicScalars.constructor(dims[2] * dims[0]);
+        let id = 0;
         for (let k = 0; k < dims[2]; k++) {
           for (let i = 0; i < dims[0]; i++) {
-            scalars.push(
-              basicScalars[i + sliceOffset * dims[0] + k * dims[0] * dims[1]]
-            );
+            id = i * dims[2] + k;
```
Comment (floryst) [inline Sources/Rendering/OpenGL/ImageMapper/index.js]: Same as https://github.com/Kitware/vtk-js/pull/606/commits/aa1d5dd138ca04b51892c42241a5ecf436678fad#r170292540

**Comment by agirault** [comment]
Ok everything is ready. Who dares review is welcome to.

@martinken the deltas are 5.76% and 5% for those two failing tests. What would you recommend I do?

**Comment by jourdain** [comment]
Did you fix the test (baseline/threshold/?) ?

Should we merge?

**Comment by agirault** [comment]
@jourdain now yes :)

**Comment by martinken** [comment]
+1

**Comment by jourdain** [comment]
Once green, I'll merge

**Comment by agirault** [comment]
Rebased on https://github.com/Kitware/vtk-js/pull/607 from @thewtex to ensure release build passes.

File: Sources/IO/Geometry/STLReader/index.js (hunk 3 above)
Comment (agirault) [reply Sources/IO/Geometry/STLReader/index.js]: @jourdain when mat4 is a `Float32Array`, `Array.isArray(mat4x4)` returns `false` in MatrixBuilder (but no warning or error message)
https://github.com/Kitware/vtk-js/commit/d1fb6679af8edfeffd4644c46e3246e09ee2abe8#diff-915eb56ed796b4179f838589e323b110R111

Which part should I modify? `MatrixBuilder.setMatrix` or this?

**Pull request merged**